contrib/cf-diag.sh

```diff
@@ -0,0 +1,240 @@
+
+# I recommend custom checks be added only at the end.
+# Skip ahead to CUSTOM TESTS.
+
+# Show host key (digest)
+printf '# '
+cf-key -p /var/cfengine/ppkeys/localhost.pub 2>/dev/null || echo SHA not available
+
+# VERSION CHECKS
+# makes use of 'ok_versions' defined at top of script
+
+version_data="$(cf-agent -V 2>/dev/null)"
+if [ "$?" = 127 ]; then # Don't go further if cf-agent missing
+  bail "$(cf-agent -V 2>&1 1>/dev/null)"
+fi
+
+# Beware literal newlines; don't try adding trailing comments!
+core_version_line="${version_data%
+*}" # e.g. 'CFEngine Core 3.7.3'
+core_version="${core_version_line##* }" # e.g. '3.7.3'
+enterprise_version_line="${version_data#*
+}" # e.g. 'CFEngine Enterprise 3.7.3'
+enterprise_version="${enterprise_version_line##* }" # e.g. '3.7.3'
+
+desc="Expected Core version"
+if printf '%s\n' "$core_version" | grep -qFx $ok_versions; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+desc="Expected Enterprise version"
+if printf '%s\n' "$enterprise_version" | grep -qFx $ok_versions; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+desc="Core version matches Enterprise version"
+# In pathological cases these may not match.
+# The individual checks for Core and Enterprise versions ok
+# may not detect these pathological cases if e.g. 3.7.3 and 3.10.2
+# are both allowed ("okay") versions, so we test it separately.
+if [ "$core_version" = "$enterprise_version" ]; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+# Show actual version data
+# (Can be commented out if you only allow one version anyway.)
+printf '# %s\n' "$core_version_line" "$enterprise_version_line"
+
+# CFENGINE FUNCTIONING
+
+# Check all three daemons as separate tests
+for d in cf-execd cf-monitord cf-serverd; do
+  desc="$d running"
+  if pgrep "$d" >/dev/null 2>/dev/null; then
+    succeed "$desc"
+  else
+    fail "$desc"
+  fi
+done
+
+# Test for recent run:
+
+# 15 minutes is three times the out-of-the-box default run interval (5 min)
+# However, as noted by Neil Watson
+# (in this ticket: https://tracker.mender.io/browse/CFE-769),
+# most large CFEngine installations use longer (less frequent) run intervals.
+
+# Note also THIS TEST MAY FAIL ON NON-LINUX SYSTEMS.
+# There is no portable approach I can find for this 'stat' command.
+# However, I consider this a minor (negligible) issue.
+# (It will also fail with coreutils 8.6, but 8.5 and 8.7+ are fine:
+# see https://unix.stackexchange.com/q/3348/135943.)  :)
+
+desc="cf-agent ran in last 15 minutes"
+if [ "$(($(date +%s)-$(stat -c %Y /var/cfengine/promise_summary.log)))" -lt 900 ]; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+# HUB CHECKS
+
+desc="Bootstrapped to hub (policy_server.dat exists)"
+if [ -f /var/cfengine/policy_server.dat ]; then
+  succeed "$desc"
+  hub_ip="$(cat /var/cfengine/policy_server.dat)"
+  printf '# Hub IP is %s\n' "$hub_ip"
+else
+  fail "$desc"
+  bail 'Not connected to any hub.'
+fi
+
+desc="Hub is connecting to host"
```

Comment:
This one is most useful in the context where you have an enterprise agent capable of serving reports. Also, it's not useful if the host is doing call collect/client initiated reporting, instead of the default hub initiated collection.

In the MPF there is augments support for setting the `client_initiated_reporting_enabled` class and the `control_server_call_collect_interval` variable. It might be useful to check/report these in the same place.
`cf-promises -c --show-classes` and `cf-promises -c --show-vars` (note in 3.12 these can take regex values)

Reply:
Interesting.  But, any check for those variables would ignore all pathological cases: host has no policy at all, host has invalid policy, host has a different policy from the hub (since the hub would have to be told not to initiate the reporting checks, something that's only visible from the host if you assume they're running the same policy).

Note that in the sequence written, this check comes long before the check for the *existence* of promises.cf - not even mentioning that promises.cf would have to be runnable for cf-promises to work on it, and also not mentioning that such a test would have to assume the standard MPF is more or less in use.

If call collect is in use, the operators in that environment would presumably be aware of that and would simply ignore the failing test or add a SKIP exception for that test.